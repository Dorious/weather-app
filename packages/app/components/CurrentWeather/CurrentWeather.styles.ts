import styled from 'styled-components';

import Symbol from '../Symbol';

export const Container = styled.header`
  padding: 1rem 0;
  text-align: center;
`;

export const Coords = styled.div`
  font-size: 1rem;
`;

export const City = styled.h1`
  font-weight: normal;
  font-size: 3rem;
`;
export const Country = styled.h2`
  font-weight: normal;
  font-size: 1rem;
  margin-top: -2rem;
  margin-bottom: 2rem;
`;
export const Sky = styled.h3``;

export const TemperatureContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const Temperature = styled.span`
  font-size: 5rem;
  padding: 0;
`;

export const FromTo = styled.p`
  font-size: 1rem;
  padding: 0;
`;

export const WeatherSymbol = styled(Symbol)`
  width: 6rem;
  height: 6rem;
`;
