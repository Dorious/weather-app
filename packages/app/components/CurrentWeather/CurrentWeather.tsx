import { memo, useMemo } from 'react';
import { decimalToSexagesimal } from 'geolib';

import { translateSymbolCode } from '../../utils/geolocation';
import { AppPosition, WeatherData } from '../../context/app.types';

import * as S from './CurrentWeather.styles';

const DEFAULT_EMPTY = '--';

export type CurrentWeatherProps = {
  geolocation?: AppPosition;
  weatherData?: WeatherData;
};

export default memo(function CurrentWeather({
  geolocation,
  weatherData,
}: CurrentWeatherProps): JSX.Element {
  const { coords = null } = (geolocation as GeolocationPosition) || {};

  const city = (weatherData?.location.components as any)?.city || DEFAULT_EMPTY;
  const country = (weatherData?.location.components as any)?.country || '';
  const lat = coords?.latitude
    ? decimalToSexagesimal(coords.latitude) +
      (coords.latitude > 0 ? ' N' : ' S')
    : DEFAULT_EMPTY;
  const long = coords?.longitude
    ? decimalToSexagesimal(coords?.longitude) +
      (coords.longitude > 0 ? ' E' : ' W')
    : DEFAULT_EMPTY;

  const timeseries = weatherData?.properties.timeseries || [];
  const current = timeseries[0];
  const airTemperature = Math.round(
    current?.data.instant.details.air_temperature
  );
  const symbolCode = current?.data.next_1_hours.summary.symbol_code;

  const [from, to] = useMemo<[number, number]>(() => {
    let from, to;

    if (timeseries) {
      for (let i = 0; i < 23; i += 1) {
        const { air_temperature } = timeseries[i]?.data.instant.details || {};
        if (!from || air_temperature > from) from = air_temperature;
        if (!to || air_temperature < to) to = air_temperature;
      }
    }

    return [from, to];
  }, [timeseries]);

  // Units
  const degrees =
    weatherData?.properties.meta.units.air_temperature[0].toUpperCase() || '';

  return (
    <S.Container>
      <S.Coords>
        Lat: {lat.replace(/.([0-9]+)" ([NSWE])$/, '" $2')} Long:{' '}
        {long.replace(/.([0-9]+)" ([NSWE])$/, '" $2')}
      </S.Coords>
      <S.City>{city}</S.City>
      <S.Country>{country || <>&nbsp;</>}</S.Country>
      <S.Sky>
        {symbolCode ? translateSymbolCode(symbolCode) : <>&nbsp;</>}
      </S.Sky>
      <S.TemperatureContainer>
        <S.WeatherSymbol code={symbolCode} />
        <S.Temperature>
          {isNaN(airTemperature) ? (
            DEFAULT_EMPTY
          ) : (
            <>
              {airTemperature}&deg;{degrees}
            </>
          )}
        </S.Temperature>
      </S.TemperatureContainer>
      <S.FromTo>
        From: {Math.round(from)}&deg;{degrees} To: {Math.round(to)}&deg;
        {degrees}
      </S.FromTo>
    </S.Container>
  );
});
