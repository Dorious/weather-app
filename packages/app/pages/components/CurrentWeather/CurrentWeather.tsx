import * as S from './CurrentWeather.styles';

export type CurrentWeatherProps = {};

export default function CurrentWeather({}: CurrentWeatherProps): JSX.Element {
  return (
    <S.Container>
      <S.Header>Gdynia</S.Header>
      <S.Sky>Partly cloudy</S.Sky>
      <S.TemperatureContainer>
        <S.WeatherSymbol code="partlycloudy_day" />
        <S.Temperature>15&deg;</S.Temperature>
      </S.TemperatureContainer>
      <S.FromTo>
        From: 15&deg; To: 15&deg;
      </S.FromTo>
    </S.Container>
  );
};