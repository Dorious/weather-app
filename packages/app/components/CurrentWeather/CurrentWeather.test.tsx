import { render } from '@testing-library/react';
import CurrentWeather from '.';

describe('CurrentWeather', () => {
  it('should render properly', () => {
    const { container } = render(<CurrentWeather />);
    expect(container).toMatchSnapshot();
  });
});
