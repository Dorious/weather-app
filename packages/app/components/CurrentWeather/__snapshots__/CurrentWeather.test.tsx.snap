// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentWeather should render properly 1`] = `
<div>
  <header
    class="CurrentWeatherstyles__Container-sc-1ecxqgh-0 cMCauJ"
  />
</div>
`;
